src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -247,29 +261,13 @@ public static unsafe SocketError TryGetAddrInfo(string name, out IPHostEntry hos
                     // We also filter based on whether IPv6 is supported on the current
                     // platform / machine.
                     //
-                    if ((pAddressInfo->ai_family == AddressFamily.InterNetwork) || // Never filter v4
-                        (pAddressInfo->ai_family == AddressFamily.InterNetworkV6 && SocketProtocolSupportPal.OSSupportsIPv6))
+                    if (pAddressInfo->ai_family == AddressFamily.InterNetwork)
                     {
-                        sockaddr = new SocketAddress(pAddressInfo->ai_family, pAddressInfo->ai_addrlen);
-                        //
-                        // Push address data into the socket address buffer
-                        //
-                        for (int d = 0; d < pAddressInfo->ai_addrlen; d++)
-                        {
-                            sockaddr[d] = *(pAddressInfo->ai_addr + d);
-                        }
-                        //
-                        // NOTE: We need an IPAddress now, the only way to create it from a
-                        //       SocketAddress is via IPEndPoint. This ought to be simpler.
-                        //
-                        if (pAddressInfo->ai_family == AddressFamily.InterNetwork)
-                        {
-                            addresses.Add(((IPEndPoint)IPEndPointStatics.Any.Create(sockaddr)).Address);
-                        }
-                        else
-                        {
-                            addresses.Add(((IPEndPoint)IPEndPointStatics.IPv6Any.Create(sockaddr)).Address);
-                        }
+                        addresses.Add(CreateIPv4Address(pAddressInfo->ai_addr, pAddressInfo->ai_addrlen));
+                    }
+                    else if (pAddressInfo->ai_family == AddressFamily.InterNetworkV6 && SocketProtocolSupportPal.OSSupportsIPv6)
+                    {
+                        addresses.Add(CreateIPv6Address(pAddressInfo->ai_addr, pAddressInfo->ai_addrlen));
```

Comment:
What if CreateIPv4/6Address return null?  If that's not possible based on construction here, it seems those checks should be moved out of these methods to the call sites.  Then this call site can remain as-is, and the other call site would do the length check at that call site.